### PR TITLE
Add polling with overhead to python sdk

### PR DESCRIPTION
**Pull request by NRHelmi**
**EDITED:** This PR adds support to polling with overhead the the python sdk.
 Polling delay is `x%` overhead of the time the transaction has been running so far.
 ~~By default the overhead is `10%`~~ no default value is set
 for transactions async the overhead is `20%` similar to what we have in golang sdk

### CODE REVIEW

File: railib/api.py [hunk 1]
```diff
@@ -325,6 +326,38 @@ def _parse_arrow_results(files: List[TransactionAsyncFile]):
...
+# polling with specified overhead
+# delay is the overhead % of the time the transaction has been running so far
+
+
+def poll_with_specified_overhead(
+    f,
+    overhead_rate: float = 0.1,
+    start_time: datetime = datetime.now(timezone.utc),
+    timeout: int = None,
+    max_retries: int = None
```
Comment (larf311) [inline railib/api.py:338]: `max_retries` is never used by any tests or examples. Speaking of which, why do we need `max_retries` when we have `timeout`?

File: railib/api.py [hunk 2]
```diff
@@ -325,6 +326,38 @@ def _parse_arrow_results(files: List[TransactionAsyncFile]):
             results.append({"relationId": file.name, "table": table})
     return results
 
+# polling with specified overhead
+# delay is the overhead % of the time the transaction has been running so far
+
+
+def poll_with_specified_overhead(
+    f,
+    overhead_rate: float = 0.1,
```
Comment (larf311) [inline railib/api.py:335]: Is 10% the standard for transactions? Seems a bit high.

File: railib/api.py (hunk 1 above)
Comment (NRHelmi) [reply railib/api.py:338]: this was inspired from the julia sdk https://github.com/RelationalAI/rai-sdk-julia/blob/main/src/api.jl#L126-L157
`timeout` and 'max_retries` could/should co-exist a polling library for python do the same
https://github.com/justiniso/polling/blob/master/polling.py#L44

File: railib/api.py (hunk 2 above)
Comment (NRHelmi) [reply railib/api.py:335]: it's `10%` for julia sdk https://github.com/RelationalAI/rai-sdk-julia/blob/main/src/api.jl#L91

File: railib/api.py (hunk 1 above)
Comment (larf311) [reply railib/api.py:338]: Then add tests. Also, 'retry' is usually something you do when you get an error. `max_tries` from the polling library seems more correct.

File: railib/api.py [hunk 3]
```diff
@@ -325,6 +326,38 @@ def _parse_arrow_results(files: List[TransactionAsyncFile]):
...
+            raise Exception(f'max retries {max_retries} exhausted')
+
+        if max_time is not None and time.time() >= max_time:
+            raise Exception(f'timed out after {timeout} seconds')
+
+        if f():
+            break
+
+        retries += 1
+        duration = (datetime.now(timezone.utc) - start_time).total_seconds() * overhead_rate
```
Comment (larf311) [inline railib/api.py:354]: Why do we use `time.time()` above and `datetime.now()` here?

Comment (NRHelmi) [reply railib/api.py:354]: we use `datetime` to convert transaction `created_on` timestamp to `utc datetime` (the comparison is based on utc timezone) which will require extra processing if using `time.time()`
we simply use `time.time()` to check timeout, we can also use `datetime`

**Review by larf311 — APPROVED** [review]

**Pull request merged**